### ISSUE TITLE
Que jobs running concurrently have interference on Rails 5.

### ISSUE THREAD

**Issue by bramski** [chanks/que#166]
Hello,

We're having a consistently repeatable problem over here of what we think is a fundamental problem of the MVCC advisory lock model in use here.

I see this code, and it doesn't seem to be effective:
https://github.com/chanks/que/blob/master/lib/que/job.rb#L88

Screenshot of our logs:
![as_a_user_i_want_lucy_to_deal_with_facebook_sending_the_same_message_twice_so_that_i_don_t_have_a_confused_conversation__-_lucy_-_confluence](https://cloud.githubusercontent.com/assets/421781/17109919/efd08cb6-524e-11e6-91eb-86a60ec528d0.png)

Our Job Runner Looks like:

```
require 'que'

module ActiveJob
  module QueueAdapters
    class ImhrJobRunner < ::Que::Job
      # Places the que job in job_data["que_ref"]
      # which can be picked up by ApplicationJob
      def run(job_data)
        @job_data = job_data

        if que_job_done? # we only need to check cause Que picks up destroyed jobs.
          log_rerun_issue_info
        else
          job_data["que_ref"] = self
          ApplicationJob.execute job_data
        end

        mark_que_job_as_done # if you reach this point it's done
      end

      public :destroy

      private

      attr_reader :job_data

      def log_rerun_issue_info
        Rails.logger.warn "#{self.class.name}: Que tried to rerun #{job_data["job_class"]}: " \
                          "#{job_data["job_id"]}"
      end

      def que_job_done?
        Rails.cache.exist?(que_job_cache_key)
      end

      def mark_que_job_as_done
        Rails.cache.write(que_job_cache_key, true, expires_in: 10.seconds)
      end

      def que_job_cache_key
        ["que-job-done", job_data["job_id"]].join("-")
      end
    end
  end
end
```

We've added Rails cache which uses redis to ensure that the jobs aren't being run twice.

**Comment by bramski** [comment]
Does the ```WARNING: you don't own a lock type of ExclusiveLock``` have something to do with this?

**Comment by chanks** [comment]
That warning message would come up if you try to release an advisory lock that hasn't been taken by that connection (it might also come up in other cases). Are you using pgbouncer, or some other connection pool outside the app? That may cause that message. Or, are you doing anything unusual with your PG connections?

Are you sure the job is being destroyed after it is run? Que tries to destroy jobs on its own that you haven't explicitly called destroy on, but it uses a `@destroyed` instance variable in the job instance itself to determine whether that is necessary, and your logic may be writing to that variable? Or possibly you're calling `destroy` in a savepoint that's getting rolled back while the main transaction continues to completion? The fact that you've marked `destroy` as public is a little suspicious to me, but I don't know what your actual job logic looks like.

**Comment by bramski** [comment]
It's just Que & AR.  No pgbouncer.
Destroy was made public so that we can put it inside the transaction explicitly and expose it to the AJ job.
The transaction isn't cancelling or getting rolled back.  In fact in the log you're seeing there is no transaction on the job which is repeating.  Adding a transaction to that job made it less consistently repeatable, but still repeatable.

**Comment by bramski** [comment]
The job is definitely destroyed.  We added an AR model to check it's existence when being run and it's typically not there.  About 80%, but it's still possible that the AR connection thinks the job is there.  I'm guessing due to the same MVCC problem that the code states.  Redis-style locking manages to be consistent here.

**Comment by joevandyk** [comment]
Is this running in Rails production mode?

**Comment by bramski** [comment]
Seeing it in ```staging``` and ```production```.

**Comment by bramski** [comment]
We've tuned the que worker to be more rapidly responsive.  We were seeing this rather intermittently and then tuned the worker wake interval to ```0.01``` and started seeing this much more consistently.

**Comment by chanks** [comment]
That's not a ridiculous wake interval. I still suspect that there's something unusual about your setup, or perhaps your Postgres configuration? Have you changed the default transaction isolation, or are you somehow running the workers themselves inside of a transaction?

I doubt that there's anything fundamentally unexpected about advisory locks and how they interact with MVCC, or it would have come up sometime in the last three years. Of course, if you had a self-contained reproduction that demonstrated the issue, that would be helpful.

**Comment by chanks** [comment]
Also, I don't know what's causing those "you don't own a lock type of ExclusiveLock" warnings, those aren't a normal thing and are a signifier that something is wrong.

**Comment by bramski** [comment]
Postgres is vanilla heroku.
Version:
``` PostgreSQL 9.5.2 on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu/Linaro 4.6.3-1ubuntu5) 4.6.3, 64-bit```

We haven't done anything to the transaction isolation.
Procfile:
```
web: bundle exec puma -C config/puma.rb
worker: bundle exec que --worker-count $WORKER_QUE_WORKERS --wake-interval $WORKER_QUE_WAKE_INTERVAL ./config/environment.rb
```

**Comment by bramski** [comment]
Show all from heroku postgres:
```allow_system_table_mods	off	Allows modifications of the structure of system tables.
application_name	Postico 1.0.8	Sets the application name to be reported in statistics and logs.
archive_command	test -f /etc/postgresql/wal-e.d/ARCHIVING_OFF || envdir /etc/postgresql/wal-e.d/env wal-e wal-push %p	Sets the shell command that will be called to archive a WAL file.
archive_mode	on	Allows archiving of WAL files using archive_command.
archive_timeout	1min	Forces a switch to the next xlog file if a new file has not been started within N seconds.
array_nulls	on	Enable input of NULL elements in arrays.
authentication_timeout	1min	Sets the maximum allowed time to complete client authentication.
autovacuum	on	Starts the autovacuum subprocess.
autovacuum_analyze_scale_factor	0.1	Number of tuple inserts, updates, or deletes prior to analyze as a fraction of reltuples.
autovacuum_analyze_threshold	50	Minimum number of tuple inserts, updates, or deletes prior to analyze.
autovacuum_freeze_max_age	200000000	Age at which to autovacuum a table to prevent transaction ID wraparound.
autovacuum_max_workers	3	Sets the maximum number of simultaneously running autovacuum worker processes.
autovacuum_multixact_freeze_max_age	400000000	Multixact age at which to autovacuum a table to prevent multixact wraparound.
autovacuum_naptime	1min	Time to sleep between autovacuum runs.
autovacuum_vacuum_cost_delay	20ms	Vacuum cost delay in milliseconds, for autovacuum.
autovacuum_vacuum_cost_limit	-1	Vacuum cost amount available before napping, for autovacuum.
autovacuum_vacuum_scale_factor	0.2	Number of tuple updates or deletes prior to vacuum as a fraction of reltuples.
autovacuum_vacuum_threshold	50	Minimum number of tuple updates or deletes prior to vacuum.
autovacuum_work_mem	-1	Sets the maximum memory to be used by each autovacuum worker process.
backslash_quote	safe_encoding	Sets whether "\'" is allowed in string literals.
bgwriter_delay	200ms	Background writer sleep time between rounds.
bgwriter_lru_maxpages	100	Background writer maximum number of LRU pages to flush per round.
bgwriter_lru_multiplier	2	Multiple of the average buffer usage to free per round.
block_size	8192	Shows the size of a disk block.
bonjour	off	Enables advertising the server via Bonjour.
bonjour_name		Sets the Bonjour service name.
bytea_output	escape	Sets the output format for bytea.
check_function_bodies	on	Check function bodies during CREATE FUNCTION.
checkpoint_completion_target	0.7	Time spent flushing dirty buffers during checkpoint, as fraction of checkpoint interval.
checkpoint_timeout	10min	Sets the maximum time between automatic WAL checkpoints.
checkpoint_warning	30s	Enables warnings if checkpoint segments are filled more frequently than this.
client_encoding	UNICODE	Sets the client's character set encoding.
client_min_messages	notice	Sets the message levels that are sent to the client.
cluster_name		Sets the name of the cluster, which is included in the process title.
commit_delay	0	Sets the delay in microseconds between transaction commit and flushing WAL to disk.
commit_siblings	5	Sets the minimum concurrent open transactions before performing commit_delay.
constraint_exclusion	partition	Enables the planner to use constraints to optimize queries.
cpu_index_tuple_cost	0.001	Sets the planner's estimate of the cost of processing each index entry during an index scan.
cpu_operator_cost	0.0005	Sets the planner's estimate of the cost of processing each operator or function call.
cpu_tuple_cost	0.003	Sets the planner's estimate of the cost of processing each tuple (row).
cursor_tuple_fraction	0.1	Sets the planner's estimate of the fraction of a cursor's rows that will be retrieved.
data_checksums	on	Shows whether data checksums are turned on for this cluster.
DateStyle	ISO, MDY	Sets the display format for date and time values.
db_user_namespace	off	Enables per-database user names.
deadlock_timeout	1s	Sets the time to wait on a lock before checking for deadlock.
debug_assertions	off	Shows whether the running server has assertion checks enabled.
debug_pretty_print	on	Indents parse and plan tree displays.
debug_print_parse	off	Logs each query's parse tree.
debug_print_plan	off	Logs each query's execution plan.
debug_print_rewritten	off	Logs each query's rewritten parse tree.
default_statistics_target	100	Sets the default statistics target.
default_tablespace		Sets the default tablespace to create tables and indexes in.
default_text_search_config	pg_catalog.english	Sets default text search configuration.
default_transaction_deferrable	off	Sets the default deferrable status of new transactions.
default_transaction_isolation	read committed	Sets the transaction isolation level of each new transaction.
default_transaction_read_only	off	Sets the default read-only status of new transactions.
default_with_oids	off	Create new tables with OIDs by default.
dynamic_shared_memory_type	sysv	Selects the dynamic shared memory implementation used.
effective_cache_size	10800000kB	Sets the planner's assumption about the size of the disk cache.
effective_io_concurrency	1	Number of simultaneous requests that can be handled efficiently by the disk subsystem.
enable_bitmapscan	on	Enables the planner's use of bitmap-scan plans.
enable_hashagg	on	Enables the planner's use of hashed aggregation plans.
enable_hashjoin	on	Enables the planner's use of hash join plans.
enable_indexonlyscan	on	Enables the planner's use of index-only-scan plans.
enable_indexscan	on	Enables the planner's use of index-scan plans.
enable_material	on	Enables the planner's use of materialization.
enable_mergejoin	on	Enables the planner's use of merge join plans.
enable_nestloop	on	Enables the planner's use of nested-loop join plans.
enable_seqscan	on	Enables the planner's use of sequential-scan plans.
enable_sort	on	Enables the planner's use of explicit sort steps.
enable_tidscan	on	Enables the planner's use of TID scan plans.
escape_string_warning	on	Warn about backslash escapes in ordinary string literals.
event_source	PostgreSQL	Sets the application name used to identify PostgreSQL messages in the event log.
exit_on_error	off	Terminate session on any error.
extra_float_digits	0	Sets the number of digits displayed for floating-point values.
from_collapse_limit	8	Sets the FROM-list size beyond which subqueries are not collapsed.
fsync	on	Forces synchronization of updates to disk.
full_page_writes	on	Writes full pages to WAL when first modified after a checkpoint.
geqo	on	Enables genetic query optimization.
geqo_effort	5	GEQO: effort is used to set the default for other GEQO parameters.
geqo_generations	0	GEQO: number of iterations of the algorithm.
geqo_pool_size	0	GEQO: number of individuals in the population.
geqo_seed	0	GEQO: seed for random path selection.
geqo_selection_bias	2	GEQO: selective pressure within the population.
geqo_threshold	12	Sets the threshold of FROM items beyond which GEQO is used.
gin_fuzzy_search_limit	0	Sets the maximum allowed result for exact search by GIN.
gin_pending_list_limit	4MB	Sets the maximum size of the pending list for GIN index.
hot_standby	on	Allows connections and queries during recovery.
hot_standby_feedback	on	Allows feedback from a hot standby to the primary that will avoid query conflicts.
huge_pages	try	Use of huge pages on Linux.
ignore_checksum_failure	off	Continues processing after a checksum failure.
ignore_system_indexes	off	Disables reading from system indexes.
integer_datetimes	on	Datetimes are integer based.
IntervalStyle	postgres	Sets the display format for interval values.
join_collapse_limit	8	Sets the FROM-list size beyond which JOIN constructs are not flattened.
krb_caseins_users	off	Sets whether Kerberos and GSSAPI user names should be treated as case-insensitive.
lc_collate	en_US.UTF-8	Shows the collation order locale.
lc_ctype	en_US.UTF-8	Shows the character classification and case conversion locale.
lc_messages	en_US.UTF-8	Sets the language in which messages are displayed.
lc_monetary	en_US.UTF-8	Sets the locale for formatting monetary amounts.
lc_numeric	en_US.UTF-8	Sets the locale for formatting numbers.
lc_time	en_US.UTF-8	Sets the locale for formatting date and time values.
listen_addresses	*	Sets the host name or IP address(es) to listen to.
lo_compat_privileges	off	Enables backward compatibility mode for privilege checks on large objects.
local_preload_libraries	pgextwlist	Lists unprivileged shared libraries to preload into each backend.
lock_timeout	0	Sets the maximum allowed duration of any wait for a lock.
log_autovacuum_min_duration	30s	Sets the minimum execution time above which autovacuum actions will be logged.
log_checkpoints	on	Logs each checkpoint.
log_connections	on	Logs each successful connection.
log_destination	syslog	Sets the destination for server log output.
log_disconnections	off	Logs end of a session, including duration.
log_duration	off	Logs the duration of each completed SQL statement.
log_error_verbosity	default	Sets the verbosity of logged messages.
log_executor_stats	off	Writes executor performance statistics to the server log.
log_file_mode	0600	Sets the file permissions for log files.
log_hostname	off	Logs the host name in the connection logs.
log_line_prefix	%m %p %u [BLACK] 	Controls information prefixed to each log line.
log_lock_waits	on	Logs long lock waits.
log_min_duration_statement	2s	Sets the minimum execution time above which statements will be logged.
log_min_error_statement	error	Causes all statements generating error at or above this level to be logged.
log_min_messages	notice	Sets the message levels that are logged.
log_parser_stats	off	Writes parser performance statistics to the server log.
log_planner_stats	off	Writes planner performance statistics to the server log.
log_replication_commands	off	Logs each replication command.
log_rotation_age	1d	Automatic log file rotation will occur after N minutes.
log_rotation_size	10MB	Automatic log file rotation will occur after N kilobytes.
log_statement	ddl	Sets the type of statements logged.
log_statement_stats	off	Writes cumulative performance statistics to the server log.
log_temp_files	10MB	Log the use of temporary files larger than this number of kilobytes.
log_timezone	UTC	Sets the time zone to use in log messages.
log_truncate_on_rotation	off	Truncate existing log files of same name during log rotation.
logfebe.identity	9b6f0c09-95fa-4e93-ade5-cfc7714d04c5	The identity of the installation of PostgreSQL.
logfebe.unix_socket	/tmp/pg_logplexcollector/pg_logplexcollector.sock	Unix socket to send logs to in FEBE frames.
logging_collector	on	Start a subprocess to capture stderr output and/or csvlogs into log files.
maintenance_work_mem	1700MB	Sets the maximum memory to be used for maintenance operations.
max_connections	500	Sets the maximum number of concurrent connections.
max_files_per_process	1000	Sets the maximum number of simultaneously open files for each server process.
max_function_args	100	Shows the maximum number of function arguments.
max_identifier_length	63	Shows the maximum identifier length.
max_index_keys	32	Shows the maximum number of index keys.
max_locks_per_transaction	64	Sets the maximum number of locks per transaction.
max_pred_locks_per_transaction	64	Sets the maximum number of predicate locks per transaction.
max_prepared_transactions	0	Sets the maximum number of simultaneously prepared transactions.
max_replication_slots	0	Sets the maximum number of simultaneously defined replication slots.
max_stack_depth	2MB	Sets the maximum stack depth, in kilobytes.
max_standby_archive_delay	-1	Sets the maximum delay before canceling queries when a hot standby server is processing archived WAL data.
max_standby_streaming_delay	-1	Sets the maximum delay before canceling queries when a hot standby server is processing streamed WAL data.
max_wal_senders	20	Sets the maximum number of simultaneously running WAL sender processes.
max_wal_size	2GB	Sets the WAL size that triggers a checkpoint.
max_worker_processes	8	Maximum number of concurrent worker processes.
min_wal_size	80MB	Sets the minimum size to shrink the WAL to.
operator_precedence_warning	off	Emit a warning for constructs that changed meaning since PostgreSQL 9.4.
password_encryption	on	Encrypt passwords.
pg_stat_statements.max	5000	Sets the maximum number of statements tracked by pg_stat_statements.
pg_stat_statements.save	on	Save pg_stat_statements statistics across server shutdowns.
pg_stat_statements.track	top	Selects which statements are tracked by pg_stat_statements.
pg_stat_statements.track_utility	on	Selects whether utility commands are tracked by pg_stat_statements.
port	5432	Sets the TCP port the server listens on.
post_auth_delay	0	Waits N seconds on connection startup after authentication.
pre_auth_delay	0	Waits N seconds on connection startup before authentication.
quote_all_identifiers	off	When generating SQL fragments, quote all identifiers.
random_page_cost	2	Sets the planner's estimate of the cost of a nonsequentially fetched disk page.
restart_after_crash	on	Reinitialize server after backend crash.
row_security	on	Enable row security.
search_path	"$user", public	Sets the schema search order for names that are not schema-qualified.
segment_size	1GB	Shows the number of pages per disk file.
seq_page_cost	1	Sets the planner's estimate of the cost of a sequentially fetched disk page.
server_encoding	UTF8	Sets the server (database) character set encoding.
server_version	9.5.2	Shows the server version.
server_version_num	90502	Shows the server version as an integer.
session_replication_role	origin	Sets the session's behavior for triggers and rewrite rules.
shared_buffers	2929MB	Sets the number of shared memory buffers used by the server.
sql_inheritance	on	Causes subtables to be included by default in various commands.
ssl	on	Enables SSL connections.
ssl_ca_file		Location of the SSL certificate authority file.
ssl_cert_file	server.crt	Location of the SSL server certificate file.
ssl_crl_file		Location of the SSL certificate revocation list file.
ssl_key_file	server.key	Location of the SSL server private key file.
ssl_prefer_server_ciphers	on	Give priority to server ciphersuite order.
standard_conforming_strings	on	Causes '...' strings to treat backslashes literally.
statement_timeout	0	Sets the maximum allowed duration of any statement.
superuser_reserved_connections	3	Sets the number of connection slots reserved for superusers.
synchronize_seqscans	on	Enable synchronized sequential scans.
synchronous_commit	local	Sets the current transaction's synchronization level.
synchronous_standby_names	follower	List of names of potential synchronous standbys.
syslog_facility	local0	Sets the syslog "facility" to be used when syslog enabled.
syslog_ident	resource13538427_heroku_com	Sets the program name used to identify PostgreSQL messages in syslog.
tcp_keepalives_count	10	Maximum number of TCP keepalive retransmits.
tcp_keepalives_idle	240	Time between issuing TCP keepalives.
tcp_keepalives_interval	30	Time between TCP keepalive retransmits.
temp_buffers	8MB	Sets the maximum number of temporary buffers used by each session.
temp_file_limit	-1	Limits the total size of all temporary files used by each session.
temp_tablespaces		Sets the tablespace(s) to use for temporary tables and sort files.
TimeZone	UTC	Sets the time zone for displaying and interpreting time stamps.
timezone_abbreviations	Default	Selects a file of time zone abbreviations.
trace_notify	off	Generates debugging output for LISTEN and NOTIFY.
trace_recovery_messages	log	Enables logging of recovery-related debugging information.
trace_sort	off	Emit information about resource usage in sorting.
track_activities	on	Collects information about executing commands.
track_activity_query_size	1024	Sets the size reserved for pg_stat_activity.query, in bytes.
track_commit_timestamp	off	Collects transaction commit time.
track_counts	on	Collects statistics on database activity.
track_functions	none	Collects function-level statistics on database activity.
track_io_timing	on	Collects timing statistics for database I/O activity.
transaction_deferrable	off	Whether to defer a read-only serializable transaction until it can be executed with no possible serialization failures.
transaction_isolation	read committed	Sets the current transaction's isolation level.
transaction_read_only	off	Sets the current transaction's read-only status.
transform_null_equals	off	Treats "expr=NULL" as "expr IS NULL".
unix_socket_group		Sets the owning group of the Unix-domain socket.
unix_socket_permissions	0777	Sets the access permissions of the Unix-domain socket.
update_process_title	on	Updates the process title to show the active SQL command.
vacuum_cost_delay	0	Vacuum cost delay in milliseconds.
vacuum_cost_limit	200	Vacuum cost amount available before napping.
vacuum_cost_page_dirty	20	Vacuum cost for a page dirtied by vacuum.
vacuum_cost_page_hit	1	Vacuum cost for a page found in the buffer cache.
vacuum_cost_page_miss	10	Vacuum cost for a page not found in the buffer cache.
vacuum_defer_cleanup_age	0	Number of transactions by which VACUUM and HOT cleanup should be deferred, if any.
vacuum_freeze_min_age	50000000	Minimum age at which VACUUM should freeze a table row.
vacuum_freeze_table_age	150000000	Age at which VACUUM should scan whole table to freeze tuples.
vacuum_multixact_freeze_min_age	5000000	Minimum age at which VACUUM should freeze a MultiXactId in a table row.
vacuum_multixact_freeze_table_age	150000000	Multixact age at which VACUUM should scan whole table to freeze tuples.
wal_block_size	8192	Shows the block size in the write ahead log.
wal_buffers	16MB	Sets the number of disk-page buffers in shared memory for WAL.
wal_compression	off	Compresses full-page writes written in WAL file.
wal_keep_segments	61	Sets the number of WAL files held for standby servers.
wal_level	hot_standby	Set the level of information written to the WAL.
wal_log_hints	off	Writes full pages to WAL when first modified after a checkpoint, even for a non-critical modifications.
wal_receiver_status_interval	10s	Sets the maximum interval between WAL receiver status reports to the primary.
wal_receiver_timeout	1min	Sets the maximum wait time to receive data from the primary.
wal_retrieve_retry_interval	5s	Sets the time to wait before retrying to retrieve WAL after a failed attempt.
wal_segment_size	16MB	Shows the number of pages per write ahead log segment.
wal_sender_timeout	1min	Sets the maximum time to wait for WAL replication.
wal_sync_method	fdatasync	Selects the method used for forcing WAL updates to disk.
wal_writer_delay	200ms	WAL writer sleep time between WAL flushes.
work_mem	100MB	Sets the maximum memory to be used for query workspaces.
xmlbinary	base64	Sets how binary values are to be encoded in XML.
xmloption	content	Sets whether XML data in implicit parsing and serialization operations is to be considered as documents or content fragments.
zero_damaged_pages	off	Continues processing past damaged page headers.```

**Comment by bramski** [comment]
Postgres Settings:  
[postgres_settings.txt](https://github.com/chanks/que/files/382273/postgres_settings.txt)

it's actually a csv.  so rename and open.  GH doesn't seem to like csv files.

**Comment by bramski** [comment]
Looks like ```read_committed``` is the default for the heroku PG instances.

**Comment by joevandyk** [comment]
Is there anything in your code/job that could be starting new database connections? Forking processes, etc? Or stopping processes?

That warning indicates that the advisory lock is being dropped and so it would make sense that Que would re-run the job.

**Comment by bramski** [comment]
You are looking at all the code pre-AJ @joevandyk .  Nothing fancy after that.  Just business logic.

**Comment by bramski** [comment]
We haven't changed the Que.mode, but we are running a separate que process.

**Comment by bramski** [comment]
Correction:
```Que.mode = :off ``` in our staging environment.

**Comment by bramski** [comment]
I can repro this consistently locally on my own postgres.  Setting concurrent worker count to 1 this doesn't repro itself.  So I suspect this is because the unlock is happening on a different connection.  I'm on rails 5.0.0 and using the AR adapter.

**Comment by joevandyk** [comment]
Hm.. I haven't tried Rails 5 yet. Wonder if 5 is doing something different?

**Comment by bramski** [comment]
WRT to the active record adapter it probably is.

**Comment by bramski** [comment]
It appears that ```work``` checks out a connection... but every internal call to execute also checks out a new connection as well.

**Comment by chanks** [comment]
Sounds like our interface to the connection pool broke somehow under Rails
5.

**Comment by bramski** [comment]
Rails 5 has taken a more predictable approach than previous comments make me think.  Everytime you checkout it gives you a different connection.

**Comment by bramski** [comment]
And the pool checks out new connections for different threads; but then will give whichever connection is available to your thread next you checkout.  So the more threads you have, the less likely the next checkout will give you the same thread.

**Comment by bramski** [comment]
I'll do my best to refactor and issue a PR that makes sense.

**Comment by chanks** [comment]
I'm on my phone so I can't check, but the AR adapter uses with_connection,
right? Does that not do the same thing anymore? Or is it no longer
reentrant?

**Comment by bramski** [comment]
It appears to no longer be reentrant.

**Comment by bramski** [comment]
I've attempted it with ```ActiveRecord::Base.connection``` and that doesn't seem to yield the proper result either.   The connection pool reaper will recycle that connection if your worker does blocking IO.

**Comment by bramski** [comment]
Ensuring that the worker loop uses the same connection seems paramount to actually executing the advisory locks.  Why not refactor the code in such a way that ensures that behavior?

**Comment by joevandyk** [comment]
Might be worth it somehow to setup a concurrent integration test.

**Comment by bramski** [comment]
https://github.com/chanks/que/pull/167 should fix this.  Not my favorite fix; but it avoids doing a major refactor that would prevent all of the re-entrant checkouts that are currently done on the adapter.

**Comment by bramski** [comment]
There are other problems in master that prevent that fix from working well.  Internal errors cause the connections not to be cleaned up and then the pool is stuck without a way to checkout a new connection.  Please let me know if you guys will have time to fix the problems or else I will have to take up this work myself.

**Comment by chanks** [comment]
I've long since started moving apps off of ActiveRecord, and I'm not familiar with the changes that have gone into it for 5.0 (I've been relying on pull requests for Rails-specific functionality for a while now). I'm also not really willing to completely refactor the worker system to support it, when a reentrant method that yields a connection for the current thread is a much simpler API to rely on (and one that 1.0 relies even more heavily on, btw).

That said, there must be some way to get this behavior from ActiveRecord 5.0 without too much code.

**Comment by bramski** [comment]
This issue just ran my PG heroku instance out of memory.  At about 3am.  Not awesome.  Any update @chanks ?

**Comment by chanks** [comment]
Haven't had time to look into this. Input from someone who knows
ActiveRecord 5 internals would be appreciated.

**Comment by ericboehs** [comment]
I'm seeing this problem as well. I really enjoy Que + AR and hope this can be resolved. @bramski Have you had any luck?

**Comment by bramski** [comment]
Zero.  We had to write a redis workaround.  We will switch to rabbit mq.
No solutions presented here for rails 5.

**Comment by 9mm** [comment]
Damn... well this is really sad. I was really looking forward to using this on a really important project.

@bramski so your PR doesn't actually help the issue then?

**Comment by bramski** [comment]
Nope.  We moved to sidekiq.

**Comment by 9mm** [comment]
Is there a way to do super reliable exactly-once processing with sidekiq? Everything I read makes it sound amazing for performance and whatnot, but I have stuff thats absolutely critical to be ran on time, reliably (dealing with payment processing). That's why I liked the ACID guarantees of Que (and Delayed Job is too slow)

**Comment by bramski** [comment]
Not of importance to us.

**Comment by chanks** [comment]
Sidekiq Pro offers this guarantee. I've never dug that far into it's
implementation, but we use it in an app at my day job and there haven't
been problems with it that I recall.

If you're interested in a job queue with a larger community that can offer
better and easier Rails/ActiveRecord integration, and ACID guarantees
aren't important to you, it's a great choice!

**Comment by bgentry** [comment]
[The Rails docs](http://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/ConnectionPool.html#method-i-with_connection) still suggest that the `with_connection` behavior is unchanged. The only changes to that code I see are from @thedarkone but I don't see any obvious problems that would have broken this behavior. https://github.com/rails/rails/commit/603fe20c0b8c05bc1cca8d01caadbd7060518141

**Comment by thedarkone** [comment]
Since I've been pinged: `with_connection` is reentrant.

```ruby
ActiveRecord::Base.connection_pool.with_connection do |conn_outer|
  ActiveRecord::Base.connection_pool.with_connection do |conn_inner|
    conn_outer == conn_inner # => true
    
    ActiveRecord::Base.connection_pool.with_connection do |another_conn_inner|
      conn_outer == another_conn_inner # => true
      conn_outer == conn_inner # => true
      conn_inner == another_conn_inner # => true
    end
  end
  
  ActiveRecord::Base.connection_pool.with_connection do |repeated_conn_inner|
    conn_outer == repeated_conn_inner # => true
  end
end
```

Leaving the outer-most `with_connection` doesn't guarantee that next time the pool will provide the same connection (obviously):

```ruby
ActiveRecord::Base.connection_pool.with_connection do |conn|
  @leaked_conn = conn
end

ActiveRecord::Base.connection_pool.with_connection do |conn|
  @leaked_conn == conn # => might be true or false, depends on the conn pool state/other threads etc.
end
```

Note that manually returning a conn to the pool breaks `with_connection` contract:

```ruby
ActiveRecord::Base.connection_pool.with_connection do |conn|
  # breaking with_connection contract:
  ActiveRecord::Base.connection_pool.checkin conn
  # see also pool.release_connection, pool.remove(conn), or ActiveRecord::Base.clear_active_connections!

  ActiveRecord::Base.connection_pool.with_connection do |conn_inner|
    conn_inner == conn # => might be true or false, depends on the conn pool state/other threads etc.
  end
end
```

My beast guess at what is happening: 3-rd party or app code, is manually releasing/returning the connection to the pool.

**Comment by chanks** [comment]
Hmm, that's interesting. Thanks for the reassurance, @thedarkone.

This doesn't seem to be a Que issue, so I'm going to close it. If anyone has any ideas on how to (cleanly) prevent this kind of occurrence, I'm open to hearing them, but I think it's reasonable for us to assume that `with_connection` is going to be both reentrant and hold the connection for the length of the block, since it's designed that way. Anyone else's code that breaks that contract is probably behaving irresponsibly, I would say.

**Comment by bramski** [comment]
This is reproduceable with vanilla rails 5 and que.  Why would you close
it?  The issue is not fixed.

**Comment by chanks** [comment]
Has anyone put together a reproduction with vanilla Rails 5?

**Comment by bramski** [comment]
The original issue this was linked to was vanilla AR with que.

**Comment by chanks** [comment]
And no other gems? It looks like there was something in your setup that was breaking the with_connection contract.

**Comment by thedarkone** [comment]
I know nothing about Que, @chanks can you test it? Or if I wanted to poke at this, all I need is postgres based vanilla Rails 5 app and then I just follow the `README.md`'s Installation/Usage sections, right?

**Comment by chanks** [comment]
@thedarkone yes, it should be pretty straightforward, though I haven't actually used Rails 5 myself, so it's possible something's missing.

**Comment by j15e** [comment]
Hello, I have the same issue with my Rails 5 setup.

I get the `WARNING: you don't own a lock type of ExclusiveLock` too. Running `ruby 2.3.1p112`, `rails 5.0.0.1` and `psql (PostgreSQL) 9.4.4`. Almost a vanilla project, I doubt anything is conflicting.

Really unfortunate issue because the gem seems great otherwise. 

You could use the [apparaisal](https://github.com/thoughtbot/appraisal) gem to test que specs against multiple rails versions.

**Comment by m-o-e** [comment]
I'm seeing the same WARNING here. Vanilla Rails5, single que worker.

This issue should be re-opened.
The que README should contain a clear warning that que is not compatible with Rails5.

**Comment by chanks** [comment]
Alright, issue reopened. I'm happy to accept a patch that isn't too invasive to Que's connection management.

**Comment by bgentry** [comment]
Since nothing changed in Que, it seems like this must be from some kind of change on the Rails/ActiveRecord side.

Has anybody been able to confirm that this issue is not present in a blank Rails 4.x app but *is* present in a blank Rails 5.x app?

**Comment by bgentry** [comment]
hey folks, I'd love to dig into this with some of my spare time these next couple weeks. Looking back through the issue, though, I don't see anything linking to a basic Rails app that reproduces this problem.

Can somebody who has encountered this issue please just create a brand new Rails app that has this problem and link to the Github repo with an explanation of how you trigger the errors?

**Comment by nbibler** [comment]
@bgentry: I created a minimum viable Rails 5 + Queue 0.12 app which shows the issue _very_ apparently:

https://github.com/nbibler/rails-5-que-exclusivelocking

You can see the commit history on that repo contains nothing unusual other than forcing Que to async mode in tests and avoiding the Rails `ActiveJob::TestCase` inheritance to allow Que to be used in test mode.

**Comment by thedarkone** [comment]
@nbibler thank you!

So what happens is that since Rails 5 it is expected that app servers/job queues are using an executor (`Rails.application.executor`) or a reloader (`Rails.application.reloader`) when entering "Rails application space" (my wording, not an official moniker). This is required so that Rails is able to maintain caches/connections/code reloading etc. `ActiveJob` does this automatically for `que` here: https://github.com/rails/rails/blob/93c9534/activejob/lib/active_job/queue_adapters/que_adapter.rb#L32 (via `ActiveJob::Base.execute`).

The problem is that `que` is entering Rails application space a bit earlier than expected. The solution is to declare the intent like this:

```diff
diff --git a/lib/que/adapters/active_record.rb b/lib/que/adapters/active_record.rb
index 083e1e6..6feeada 100644
--- a/lib/que/adapters/active_record.rb
+++ b/lib/que/adapters/active_record.rb
@@ -59,7 +59,9 @@ module Que
       private
 
       def checkout_activerecord_adapter(&block)
-        ::ActiveRecord::Base.connection_pool.with_connection(&block)
+        ::Rails.application.executor.wrap do
+          ::ActiveRecord::Base.connection_pool.with_connection(&block)
+        end
       end
     end
   end
```

`Que::Adapters::ActiveRecord#cleanup!`'s call into `::ActiveRecord::Base.clear_active_connections!` thus becomes probably unnecessary as Rails [does that automatically](https://github.com/rails/rails/blob/3c785bd/activerecord/lib/active_record/query_cache.rb#L39) upon leaving an outer most executor block (executor is reentrant).

For more info there is a WIP guide rails/rails#27494 in the works.

**Comment by ericboehs** [comment]
@thedarkone Nice work! Would you want to open a pull request for that?

**Comment by chanks** [comment]
Thanks thedarkone, very interesting. Tomorrow is my open source day at work
so I should be able to review all that and probably make a bugfix release.

**Comment by bgentry** [comment]
@nbibler thank you for being a good OSS citizen and providing a reproduction!

@thedarkone thanks for so quickly finding the issue afterwards and suggesting a fix :clap:

@chanks lmk if I can help, but it looks like you've got a handle on this one now 👍

**Comment by chanks** [comment]
I've just spent some time investigating this with @nbibler's reproduction (thanks @nbibler!).

Wrapping the connection checkout in `::Rails.application.executor.wrap{}` as @thedarkone suggests does eliminate the warnings, but it seems to be a suboptimal solution to me, as it assumes that a Rails application is present, which it won't necessarily be (this solution won't work for someone who's using Que with ActiveRecord but not the rest of Rails, for example).

I tried to figure out how ActiveRecord in isolation manages its execution context, because I'd like to know more about what's going on under the covers here, but the only place that I can find in ActiveRecord master that interacts with the executor is the query cache. The query cache seemed like a plausible source for this problem (it could try to cache the job lock query, which results in the same job being worked twice and Que thinking the lock has been taken when it hasn't) but I tried disabling it and the warnings continued.

`Rails.application.executor` refers to an anonymous subclass of ActiveSupport::Executor, while the query cache adds its hooks to the ActiveSupport::Executor class itself. It occurred to me that the following should solve the problem more cleanly (without a dependency on `Rails.application`), but it doesn't have any effect, and I'm not sure why yet:

```ruby
if defined?(ActiveSupport::Executor)
  ActiveSupport::Executor.wrap do
    ::ActiveRecord::Base.connection_pool.with_connection(&block)
  end
else
  ::ActiveRecord::Base.connection_pool.with_connection(&block)
end
```

I'll try to dig into this some more later, but if anyone that knows Rails' internals can point me in the right direction, it'd be much appreciated.

**Comment by thedarkone** [comment]
@chanks yes, the problem stems from `QueryCache` (see [here](https://github.com/rails/rails/blob/3c785bd/activerecord/lib/active_record/query_cache.rb#L39)), because as things stand today - it does "after request" connection clean up, where "after request" is the executor's outermost block. In terms of `que`'s code base, right now the executor's outermost block ends right after [this line](https://github.com/chanks/que/blob/master/lib/que/job.rb#L17) (upon the return from `run(*attrs[:args])` the current thread's connection was "cleaned up" == returned to the pool).

Up until rails/rails@d3c9d808e3e conn clean up was a responsibility of a separate rails middleware, with that commit this functionality was merged into `QueryCache` (yes, this is wrong separation of concerns, but things would still be broken even if it was kept separate, as surely that code would still be injected into executor life-cycle and `que` would still run into trouble 😞). Disabling `QueryCache` doesn't help, because the code in question still runs (yes, this might be a bug, but the right fix is to move the connection release into its on executor callback, but that still wouldn't help `que`).

As to how properly handle running `que` with vanilla `active_record`: `QueryCache`'s executor callback is installed via a railtie https://github.com/rails/rails/blob/5473e39/activerecord/lib/active_record/railtie.rb#L161, so if run outside of full Rails stack, the code that breaks `que` is not used. Furthermore `ActiveJob` hooks into Rails executor via a railtie as well https://github.com/rails/rails/blob/master/activejob/lib/active_job/railtie.rb#L25. In other words, outside of Rails (1) executor doesn't have the clean up code and (2) active job does not hook up into executor.

The correct way to handle this is probably via:
```ruby
if defined?(Rails)
  Rails.application.executor.wrap do
    ::ActiveRecord::Base.connection_pool.with_connection(&block)
  end
else
  ::ActiveRecord::Base.connection_pool.with_connection(&block)
end
```

Or `que`'s own railtie could set a "running in Rails, better use an executor" flag (since if `que`'s railtie is run, `QueryCache` railtie will also run etc).

**Comment by nbibler** [comment]
Presuming we'd like to keep Rails 4 compatibility in the library, then perhaps doing something like:

```ruby
if defined?(Rails) && defined?(Rails.application.executor)
  Rails.application.executor.wrap do
    ::ActiveRecord::Base.connection_pool.with_connection(&block)
  end
else
  ::ActiveRecord::Base.connection_pool.with_connection(&block)
end
```

Would allow existing behavior to continue in Rails 4 and ActiveRecord-only environments. But, if `Rails.application.executor` exists, it'll instead jump to the Rails 5-compatibility mode and use it.

**Comment by chanks** [comment]
Alright, I just applied that fix and pushed 0.12.1. Thanks everyone for your help!

**Comment by nbibler** [comment]
Thanks everyone! The 0.12.1 release appears to fix the issue in the test repo.

**Comment by wtfiwtz** [comment]
I'm getting a follow-up issue to this with Que 0.13.1 and `Que.mode = :async` - seems to be related to Rails 5.1.3 connection management in a recently upgraded app (4.2.6 => 5.1.3). Basically `rails server`, `rails console` and `rspec` can often deadlock, and getting these strange errors sometimes too:
https://www.postgresql.org/message-id/7205.1262205377%40sss.pgh.pa.us

The solution seems to be switch to `:sync` mode in the short-term.

I'm looking into this a bit further when I have a chance...

**Comment by chanks** [comment]
Hmm, I wonder if this Rails.application.executor API changed in 5.1...

**Comment by chanks** [comment]
I updated @nbibler's example app (which replicates the problem in a test case) and I'm able to confirm that the behavior didn't exist in 5.0.5 but does in 5.1.0. While looking through the 5.1 release notes [this note](http://edgeguides.rubyonrails.org/5_1_release_notes.html#transactional-tests-with-multiple-connections) about transactional tests stood out to me as an issue, so I tried setting `self.use_transactional_tests = false` in the test class and that seems to fix it in the test app.

So I guess the lesson from that is don't use `:async` inside Rails' transactional tests, but that's a place where you should be using `:sync` anyway. I don't see any reason yet why it would fail in `rails server` or `rails console`...?

**Comment by chanks** [comment]
Alright, I was able to reproduce the issue in a console session and I'm pretty sure the issue is in our hooking into AR's transaction callbacks to wake a worker after a transaction inserting a job is committed. @wtfiwtz Would you mind pointing your Gemfile to commit `ba76522` and see if that fixes the issue for you?

**Comment by wtfiwtz** [comment]
Hi @chanks thanks for looking at this so quickly! Greatly appreciated!

It initially looked like your change did fix it, but I've run `rspec` quite a few times and the issue is still there. I've tried `pg` version 0.19 and 0.20, `rails` versions 5.0.5 and 5.1.3 on commit `ba76522` and they still have the issue.

It definitely seems to be related to the rspec transactional tests, and definitely more frequently with rails 5.1.3 than 5.0.5... I often have to restart Postgres to reset the database. I think the rails console and server deadlocks are just due to the postgres connections getting corrupted.

![image](https://user-images.githubusercontent.com/1445450/29007373-435a0c20-7b45-11e7-8db5-5da28ae86783.png)

In my most recent testing, rspec tests would succeed up till a point. The last test that succeeded gives this error: `PG::UnableToSend: insufficient data in "T" message` (in the after test hooks?) and all subsequent tests give `PG::ConnectionBad: PQsocket() can't get socket descriptor: ROLLBACK`

[rspec-log.txt](https://github.com/chanks/que/files/1203475/rspec-log.txt)

This is with PostgreSQL 9.6.2. 

Technically I don't need to use `:async` mode anymore (we've moved off Heroku onto my own Digital Ocean box now) but it would be good to figure out why it's now an issue.

I've installed `sigdump` now so hopefully I'll be able to dump a stack trace when it deadlocks again.

Maybe hold off on any further testing until I can isolate this a bit further ;)... should be able to spend some time on it later tonight...

**Comment by chanks** [comment]
My understanding from my research today is that Rails changed its behavior
inside transactional tests, which is one reason why my reproduction today
was failing. But I don't think it should be an issue for users in general,
because :async mode isn't really intended to be used inside a transactional
test anyway. In testing you should be using :off mode (if you want jobs to
simply be stored in the DB, so you can assert on their storage state) or
the :sync mode (so jobs are run immediately in the same thread, and you can
assert on their effects).

Or am I misunderstanding what you're trying to do?

**Comment by wtfiwtz** [comment]
Hi @chanks,

I think you are right - I've reviewed the changes for transactional tests and that is definitely a result of the threads sharing the DB connection during rollbacks with some sort of race condition.

https://github.com/rails/rails/pull/28083

Also some related discussions here:

https://github.com/rails/rails/issues/28466
https://github.com/rails/rails/issues/29164
https://github.com/rosenfeld/rspec_nested_transactions/issues/2

So the solution is to disable Que's `:async` mode, or disable `config.use_transactional_fixtures = false` in Rspec, or `self.use_transactional_tests = false` inside individual tests... there isn't any way to disable the new behaviour in Rails 5.1.x otherwise from what I can see. As you have said above, the `:sync` mode is more appropriate for tests anyway, so it tests your job's behaviour at the same time.

Not yet sure why it deadlocked on Rails 5.0.5 but maybe that was a one-off... 🙁 

Thanks for your help! Something to be aware of in multi-threaded Rails code from now on.

**Comment by chanks** [comment]
Yeah.

So, I fixed the issue temporarily by just commenting out the AR transaction hooks to wake a local worker after the transaction commits. I'm not sure that it's super-necessary to keep that code, since most people don't run the workers in their web processes anymore, I think, and LISTEN/NOTIFY is the better long-term solution anyway.

@joevandyk you wrote that code originally (and I think you have a better understanding of the AR transaction hooks than I do), is it still important for you and do you want to take a stab at updating it for Rails 5.1? If not I'm inclined to just remove it.

**Comment by chanks** [comment]
@wtfiwtz I just released 0.14.0 without the AR transaction hook. Thanks for bringing this to my attention!

**Comment by joevandyk** [comment]
I'd remove it.

**Comment by wtfiwtz** [comment]
Thanks @chanks! I appreciate your efforts to jump on this so quickly!
